### ISSUE TITLE
Java Module System assertions

### ISSUE THREAD

**Issue by xenoterracide** [assertj/assertj#3521]
#### Feature summary

It would be nice to be able to test java modularity.

I'm not certain what the best API is. I imagine that it has to be a completely stringy API, but I could be wrong, maybe there's a way to import the module-info.class, although for someone using an automatic module that'd have to use something stringy.

```java
assertThatModule("org.assertj.core").exports("org.assertj.core.api", "org.assertj.core...").noExportOf("org.assertj.core.something",...). noExportMatches(".*internal.*").exportedType("org.assertj.core.api.Assertions").notExportedType("..."); // note: apparently there's an option to do "exports to"

assertThatModule("org.assertj.core").opens("com.myorg.SomeClass").to("spring.core").to("...hibernate"); //note: opens doesn't always have a "to"

assertThatModule("org.assertj.core").requires("javax.sql").isStatic().requires("java.base").isTransient(); // note: I don't care about `requires` by itself so much, but would be good to enforce that an optional dependency is optional
```

active use case: https://github.com/mockito/mockito/pull/3381

**Comment by scordio** [comment]
Does this overlap with what I've mentioned at https://github.com/orgs/assertj/discussions/2229?

> Add [`Module`](https://docs.oracle.com/javase/9/docs/api/java/lang/Module.html) assertions and `module()` navigation method to `Class` assertions

I'm not entirely sure we can pick a `Module` by name, which is probably – with a brief look, `Class::getModule` seems to be the entry point for such instances.

Anyway, I like the idea and we probably have to play with what the JDK offers a bit more in order to properly shape the API 🙂

**Comment by xenoterracide** [comment]
> Anyway, I like the idea and we probably have to play with what the JDK offers a bit more in order to properly shape the API 🙂

Yeah, I just kind of typed out random possibilities that I think should be able to be asserted based on what I've done with the module-info's.

 I'm 90% interested in validating what isn't present. Since I can validate what is available by writing test suites that have their own `module-info.java`. harder (not impossible) to write a test suite that says I can't access something. essentially... I can't use `Foo.class.getModule` if `Foo.class` isn't available at compile time. I imagine there must be another way to do that, maybe from the classloader?
 
 Yes it probably overlaps with that discussion. I only skimmed it.
